12 refactor rpc client api

fkengun opened the pull request:
The RPCProxy classes are now divided into Visor version and Client version so that we can develop Visor logic independently.
The ChronoLogClient_lib is now a dynamic library for linking.
The test cases for clients are now moved to the re-organized unified root test directory.

fkengun committed:
Hide ChronoVisor functions from Client side APIs

fkengun committed:
Use dynamic linking for ChronoLogClient_lib

fkengun committed:
Move client tests to the unified root test directory
Generate server_list file after build automatically
Update README.md accordingly

ibrodkin reviewed (approved):
the client side of RPC communications is separated and the changes look good overall 

Going forward I recommend using unique_ptr for the singleton classes and making the mutex objects for the ClientRegistry and the MetadataRegistry the members of the registries they guard instead of independent global variables.  Some of the header files that are remnants of the earlier implementations  are still being referenced here and there. But these things can be taken care of later on and do not preclude this merge

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronicleMetadataRPCVisor.h:
```diff
@@ -46,49 +48,28 @@ class ChronicleMetadataRPCProxy {
         return g_chronicleMetaDirectory->create_chronicle(name);
     }
```
Can you put this extern variable in a header file or make it a member of the class? It a variable that is defined once globally, you shouldn't have to declare it locally in all functions.

aparnasasidharan2017 replied:
Do you want to consider reducing the number of LOG statements in the code? A few assertions of invariants required for correct behavior of the function should be enough.

fkengun replied:
> Can you put this extern variable in a header file or make it a member of the class? It a variable that is defined once globally, you shouldn't have to declare it locally in all functions.
I will do that.



> Do you want to consider reducing the number of LOG statements in the code? A few assertions of invariants required for correct behavior of the function should be enough.
These are LOGDs. It should not affect the performance or the readability in releases.

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronoLogAdminRPCVisor.h:
```diff
@@ -16,11 +16,12 @@
+        rpc = ChronoLog::Singleton<ChronoLogRPCFactory>::GetInstance()
+                ->GetRPC(CHRONOLOG_CONF->RPC_BASE_SERVER_PORT);
         func_prefix = "ChronoLog";
         switch (CHRONOLOG_CONF->RPC_IMPLEMENTATION) {
```
Does MetadataRPCVisor and AdminRPCVisor require separate RPC instances? Can you invoke one RPC instance and use it for both Admin and Metadata Visors? Are you creating two visors? Why not have one visor class that does both Admin and Metadata operations?

fkengun replied:
Yes, there is actually only one RPC instance. All RPCs are invoked via that instance. That is how the current Visor test case works. I just want to split the logic into separate classes so that a simplified Visor can have only admin or metadata APIs.

aparnasasidharan2017 replied:
Ok

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronoLogAdminRPCVisor.h:
```diff
@@ -50,37 +51,33 @@ class ChronoLogAdminRPCProxy {
         return true;
     }
 
-    bool Connect(const std::string &uri, std::string &client_id) {
```
Do you need a separate LocalConnect and Connect?

fkengun replied:
LocalConnect means it is local, i.e. on the Visor. Connect means it is a Client call, i.e., a remote call.

aparnasasidharan2017 replied:
And the remote Connect RPC calls LocalConnect... LocalConnect does not make sense otherwise.

aparnasasidharan2017 commented on Client/include/ChronicleMetadataRPCClient.h:
```diff
@@ -0,0 +1,110 @@
+        switch (CHRONOLOG_CONF->RPC_IMPLEMENTATION) {
+            case CHRONOLOG_THALLIUM_SOCKETS:
+            case CHRONOLOG_THALLIUM_TCP:
+            case CHRONOLOG_THALLIUM_ROCE:
```
Do you want to consider adding some error checks at the client  and server side for unsupported server configurations - protocol combinations etc? Or are we fixing a default configuration and keeping the inputs to hostname and port number?

fkengun replied:
Good idea. I will add the server configuration check. 
Yes, we have a list of default configurations in ConfigurationManager, which uses 127.0.0.1 as the default IP, 5555 as the default port, and ofi+sockets as the default protocol.

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronicleMetadataRPCVisor.h:
```diff
@@ -98,97 +79,60 @@ class ChronicleMetadataRPCProxy {
     bool LocalDestroyStory(std::string &chronicle_name, std::string &story_name, const int &flags) {
         LOGD("%s is called in PID=%d, with args: chronicle_name=%s, story_name=%s, flags=%d",
              __FUNCTION__, getpid(), chronicle_name.c_str(), story_name.c_str(), flags);
         extern std::shared_ptr<ChronicleMetaDirectory> g_chronicleMetaDirectory;
```
It may be wise to add some basic error checks in all functions that access the metadata data structure. Check for input data integrity.

fkengun replied:
Good idea. I will add that.

fkengun committed:
Replace redundant return value output with assert in test cases

fkengun committed:
Aggregate external global variable in both RPC in Visor

fkengun committed:
Only Client needs to create client-mode Thalium engine, remove redundant Visor-side client-mode engine

fkengun committed:
Add error code and basic error checking
All Client APIs return integer instead of boolean value
Instanciate RPC in ChronoLogClient constructors after configuration is set
Add a new ChronoLogClient constructor to accept protocol, ip and port to customize RPC configuration

FIXME: returning data using parameter is not working, API still returns success anyway for now

aparnasasidharan2017 commented on Client/include/ChronicleMetadataRPCClient.h:
```diff
@@ -0,0 +1,110 @@
+        for (auto iter = attrs.begin(); iter != attrs.end(); ++iter) {
+            LOGD("%s=%s", iter->first.c_str(), iter->second.c_str());
+        }
+        return CHRONOLOG_RPC_CALL_WRAPPER("CreateChronicle", 0, int, name, attrs, flags);
```
Shouldn't we check the return values of these RPCs for error codes and report them to the client? Throw an exception and terminate the client program if the RPC returns error codes or allow the client to retry with new RPC arguments. Maybe you should distinguish between these two strategies : when should the client program terminate and when should it retry?

fkengun replied:
try-catch is added at the Visor side now. Different error code in the return value will indicate the error. By this macro, the return value of RPC will be forwarded to client apps directly. Users are required for check return values before moving forward.

aparnasasidharan2017 replied:
Will you add this to the design document? Also mention places in the code where you added try-catch to terminate/abort the program vs return error code and continue execution.

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/errcode.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef CHRONOLOG_ERRCODE_H
+#define CHRONOLOG_ERRCODE_H
+
+#define CL_SUCCESS                  0
```
Please check this. It should be consistent with true/false boolean values expected by the caller. Maybe CL_SUCCESS should be 1 (true)?

fkengun replied:
In my current design, the caller should check if return value is equal to CL_SUCCESS for successful operation. All the APIs should return int. Positive numbers are reserved for returning updated ChronoTick after collision resolution in Record(). Negative numbers are left for all errors.

aparnasasidharan2017 replied:
What about functions that return boolean values and internally call functions that return error codes? Will the error code be typecast to boolean or will you add checks there?

fkengun replied:
Do you mean all Client APIs return either true (SUCCESS) or false (FAIL) to the users and users can GetErrCode() to know what exact error is when they get a false return value? That means the Client library needs to store some states internally. I want to keep it state-less. Returning error code directly as the return value does not require any state memorization in Client library.

aparnasasidharan2017 replied:
I was referring to existing functions that return boolean values and use error codes internally. You have to make sure it returns true on CL_SUCCESS a nd false otherwise if you don't throw an exception. You have to convert CL_SUCCESS which is of value 0 to boolean true in such functions. Anyway, inconsistencies in return values can be verified during testing.

fkengun replied:
I have updated the Client APIs in https://github.com/scs-lab/ChronoLog/blob/a71d7a0702e4b796d9cd5944af4683753b14627f/Client/include/client.h. They all return int now.

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -175,20 +196,24 @@ uint64_t ChronicleMetaDirectory::playback_event(uint64_t sid) {
     uint64_t cid = CityHash64(name.c_str(), name.size());
     Chronicle *pChronicle = chronicleMap_->at(cid);
-    return pChronicle->getPropertyList().at(key);
+    value = pChronicle->getPropertyList().at(key);
```
You can add an existence check here. Return SUCCESS only if chronicle exists.

fkengun replied:
done

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronoLogAdminRPCVisor.h:
```diff
@@ -37,50 +41,52 @@ class ChronoLogAdminRPCProxy {
+        }
+        if (g_clientRegistryManager) {
+            return g_clientRegistryManager->add_client_record(client_id, record);
+        }
```
What do you want to do if the clientRegistryManager does not exist? Do you want to report the error to the server program? If you are checking for nullpointers, do it consistently in all modules using assertions or throw exceptions. If there is a nullpointer in the server APIs during client RPC calls, then it is due to incorrect server initialization - throw an exception in the server program and abort/terminate server.

fkengun replied:
I did this just for debugging the segfault I had before. What we can do is to terminate Visor when it is null.

aparnasasidharan2017 replied:
ok

aparnasasidharan2017 commented on ChronoVisor/include/ChronicleMetaDirectory.h:
```diff
@@ -19,23 +19,23 @@ class ChronicleMetaDirectory {
-    std::string get_chronicle_attr(std::string &name, const std::string& key);
-    bool edit_chronicle_attr(std::string &name, const std::string& key, std::string value);
+    int get_chronicle_attr(std::string& name, const std::string& key, std::string& value);
+    int edit_chronicle_attr(std::string& name, const std::string& key, const std::string& value);
```
Do you want to duplicate the data in chronicleMap and StoryMaps and maintain acquiredChronicleMap_ and acquiredStoryMap? Can you store the acquired info in the primary data structures? Otherwise, you will have to update the primary and auxiliary data structures atomically (within the same critical section).

fkengun replied:
Currently, both ChroniMap and acquiredChronicleMap are both unordered_map<uint64_t, Chronicle *>. Their values are pointers to the same Chronicle instances. They are like two access points to the objects. We might want to maintain something else in acquiredChronicleMap, for instance, information of a Chronicle that only effective for this acquisition session. I think it is okay for now. We only need to make one change via either map. There is no primary or auxilliary data structures.

aparnasasidharan2017 replied:
Won't the values pointed to by the pointers be updated when the primary data structure inserts/deletes? Isn't ChronicleMap the primary data structure which contains all entries and the acquiredChronicleMap a subset of entries in ChronicleMap? Do you want to document this? There is potential for data races here, APIs can access the same memory location through two data structures. Make sure it is tested properly. Add some test cases to cover their usage.

fkengun replied:
Yes, ChronicleMap and acquiredChronicleMap are kinda primary/auxiliary from your point of view. I wanted to say that we do not maintain two copies of each entry. I do have a simple single-threaded test case ready which emulates this situation. At least the already-inserted entry does not change when a new entry is inserted (not overwrite the existing one). The data race could be a problem. What I believe we can do is to make sure there is no concurrent access to the same Chronicle/Story. I will update my test case to multithreaded version to test data race.

aparnasasidharan2017 replied:
Yeah, if we have a global mutex, which is the situation now, this may work. But there will be data races when you replace the stl maps and global mutex with concurrent hash tables. Maybe you should add this to our design document and we don't forget this issue when we replace the stl maps with HCL containers.

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/ChronoLogAdminRPCVisor.h:
```diff
@@ -12,15 +12,19 @@
 #include "ClientRegistryManager.h"
 
-class ChronoLogAdminRPCProxy {
+extern std::shared_ptr<ClientRegistryManager> g_clientRegistryManager;
```
Do you want to combine AdminRPC proxy and MetadataRPC proxy into a single class and invoke its instance once? I think it is safe to assume that a client calls Chronolog metadata RPCs only after it connects to a server.

fkengun replied:
I will push it for later.

fkengun replied:
I did it anyway in https://github.com/scs-lab/ChronoLog/commit/e2e76404348bbf333452eede46734ff062c2d490.

aparnasasidharan2017 replied:
ok

fkengun committed:
Catch exception of querying non-existing entry in ChronicleMap and StoryMap

fkengun committed:
Update acquisitioncount on acquire and release
Check acquisitioncount on destroy
Initialize attributes of Chronicle and Story

fkengun committed:
Fix wrong return type of acquisitioncount-related functions

fkengun committed:
Combine AdminRPC class and MetadataRPC class into one, RPCVisor for Visor and RPCClient for Client, respectively

fkengun committed:
Fix broken thallium_server

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -40,65 +37,98 @@ bool ChronicleMetaDirectory::create_chronicle(const std::string& name) {
+    Chronicle *pChronicle;
+    try {
+        pChronicle = chronicleMap_->at(cid);
+    } catch (const std::out_of_range& e) {
```
Do you want to throw an exception if the cid does not exist in the map? That would abort the server. Wouldn't you want the client to retry?

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -40,65 +37,98 @@ bool ChronicleMetaDirectory::create_chronicle(const std::string& name) {
+        if (pChronicle->getAcquisitionCount() != 0) {
+            return CL_ERR_ACQUIRED;
+        }
         delete pChronicle;
```
what if you delete the chronicle in memory and another client tries to create a chronicle with the same cid (name)? Can you update the chronicle class in memory and the map entry atomically? 
What you could do is erase the pointer from the hashtable, make sure it is unreachable from the rest of the code and delete it afterwards. But, the erase function should return the pointer for that to work. 
Or you lock(); ptr = map.find(key); map.erase(key); delete ptr; unlock();

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -40,65 +37,98 @@ bool ChronicleMetaDirectory::create_chronicle(const std::string& name) {
-    return res.second;
+    if (res.second) {
+        return CL_SUCCESS;
+    } else {
```
Oh, looks like you haven't added locks. You should add mutual exclusion to the metadirectory map accesses.

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -108,61 +138,87 @@ bool ChronicleMetaDirectory::create_story(std::string &chronicle_name,
         Chronicle *pChronicle = chronicleRecord->second;
         LOGD("Chronicle@%p", &(*pChronicle));
         bool res = pChronicle->addStory(chronicle_name, story_name, attrs);
```
What if the story already exists in the chronicle map? You have to add it only once and fix the memory leak if it is a duplicate.

fkengun committed:
Return error code in internal Chronicle functions and forward as much info about the error as possible to the users

fkengun committed:
Create a new class ClientInfo as the value field in ClientRegistry to get rid of socketpp-based ClientRegistryInfo implementation

fkengun committed:
Add missing files for combined RPC class and ClientInfo class

ibrodkin commented on Client/include/RPCClient.h:
```diff
@@ -0,0 +1,129 @@
+        LOGD("%s constructor finishes, object created@%p in thread PID=%d",
+             typeid(*this).name(), this, getpid());
+    }
+
```
Client is the public client API for ChronoLog and will contain more than just RPC Client for communicating with the Visor.  having RPCClient as a member of Client would give us more flexibility to tweak implementation without affecting the ClientAPI. I would not combine them.
Though I suggest that RPCClient does not include header files from the ChronoAPI directory. any header files that are included here will have to be exposed to the client . I suggest that anything RPCClient.h needs is defined directly in this file or eliminated.

fkengun committed:
Move files used by both ChronoVisor and Client to ChronoAPI directory
Remove redundant global_var_client.h file
Remove redundant include in CMakefiles.txt

aparnasasidharan2017 commented on ChronoAPI/ChronoLog/include/RPCVisor.h:
```diff
@@ -0,0 +1,432 @@
+            return CL_ERR_INVALID_ARG;
+        }
+        if (g_clientRegistryManager) {
+            return g_clientRegistryManager->add_client_record(client_id, record);
```
Make this function consistent with LocalDisconnect. Add a return type for all conditions (if-else), otherwise you get compilation warnings.

fkengun committed:
Add mutex to protect concurrent access to ChronicleMap, AcquiredChronicleMap, AcquiredStoryMap
Add mutex to Chronicle to protect concurrent access to its StoryMap, Stats and ArchiveMap
FIXME: A failing multithreaded test case reporting [error] ABT_xstream_self failed (ret = 4)

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -28,141 +28,342 @@ ChronicleMetaDirectory::~ChronicleMetaDirectory() {
     uint64_t cid = CityHash64(name.c_str(), name.size());
+    std::lock_guard<std::mutex> lock(g_chronicleMetaDirectoryMutex_);
+    /* Check if Chronicle already exists, fail if true */
+    if (chronicleMap_->find(cid) != chronicleMap_->end()) return CL_ERR_CHRONICLE_EXISTS;
```
delete pChronicle if it exists.

aparnasasidharan2017 commented on ChronoVisor/include/Chronicle.h:
```diff
@@ -76,94 +84,125 @@ class Chronicle {
+        pStory->setProperty(attrs);
         pStory->setSid(sid);
         pStory->setCid(cid);
+        std::lock_guard<std::mutex> lock(storyMapMutex_);
```
Why do you need this storyMapMutex_? Doesn't every chronicle object have its local story maps? The chronicle map is locked prior to this. Only one thread can modify a chronicle object.

aparnasasidharan2017 commented on ChronoVisor/include/Chronicle.h:
```diff
@@ -76,94 +84,125 @@ class Chronicle {
+        pArchive->setProperty(attrs);
         pArchive->setAid(aid);
         pArchive->setCid(cid);
+        std::lock_guard<std::mutex> lock(archiveMapMutex_);
```
Same comment about archiveMutex_. Looks okay otherwise. I'll go over it once again tomorrow and approve it once you finish testing. Can you add a link to the tests here, for future reference?

fkengun committed:
Fix memory leak in create_chronicle

fkengun committed:
Remove redundant locks when accessing StoryMap and ArchiveMap in a Chronicle

fkengun committed:
Add missing include of global_var_visor.h

aparnasasidharan2017 commented on ChronoVisor/include/Story.h:
```diff
@@ -54,16 +62,30 @@ class Story {
     void setSid(uint64_t sid) { sid_ = sid; }
     void setCid(uint64_t cid) { cid_ = cid; }
-    void setStats(const StoryStats &stats) { stats_ = stats; }
+    void setStats(const StoryStats &stats) { std::lock_guard<std::mutex> lock(statsMutex_); stats_ = stats; }
```
What is this mutex for? stats is not a global variable, and story objects are accessed only through their parent chronicle objects (protected by mutexes), do you need to add this lock here?

aparnasasidharan2017 commented on ChronoVisor/include/Story.h:
```diff
@@ -54,16 +62,30 @@ class Story {
     void setEventMap(const std::unordered_map<std::string, Event> &eventMap) { eventMap_ = eventMap; }
 
+    uint64_t incrementAcquisitionCount() {
+        std::lock_guard<std::mutex> lock(statsMutex_);
```
Same comment regarding statsMutex for increment and decrement functions

aparnasasidharan2017 commented on ChronoVisor/include/Chronicle.h:
```diff
@@ -154,16 +172,33 @@ class Chronicle {
     size_t getStoryMapSize() { return storyMap_.size(); }
     size_t getArchiveMapSize() { return archiveMap_.size(); }
 
+    std::mutex statsMutex_;
```
Do you need these three mutexes? stats, attrs, propertyList, metadataMap, storyMap and archiveMap are private members of a chronicle object. In our APIs, chronicle objects are accessed by visor threads through ChronicleMetadirectory (protected  by mutexes). All the members of a chronicle object can be accessed only through these APIs and the chronicle mutex.

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -175,20 +376,40 @@ uint64_t ChronicleMetaDirectory::playback_event(uint64_t sid) {
+    Chronicle *pChronicle;
+    std::lock_guard<std::mutex> lock(g_chronicleMetaDirectoryMutex_);
+    try {
+        pChronicle = chronicleMap_->at(cid);
```
Remove this. We can add exceptions after a meeting/discussion about error handling later.

aparnasasidharan2017 commented on ChronoVisor/src/ChronicleMetaDirectory.cpp:
```diff
@@ -28,141 +28,342 @@ ChronicleMetaDirectory::~ChronicleMetaDirectory() {
+        }
+    } else {
+        return CL_ERR_NOT_EXIST;
+    }
```
Return CL_ERR_NOT_EXIST if chronicle doesn't exist.

fkengun committed:
Remove redundant mutex in Chronicle and Story
Fix wrongly placed global_var_visor.h
